src/datasets/iterable_dataset.py

```diff
@@ -170,6 +172,85 @@ def shard_data_sources(self, shard_idx: int) -> "CyclingMultiSourcesExamplesIter
         raise NotImplementedError("Sharding a CyclingMultiSourcesExamplesIterable is not implemented")
 
 
+class VerticallyConcatenatedMultiSourcesExamplesIterable(_BaseExamplesIterable):
+    def __init__(self, ex_iterables: List[_BaseExamplesIterable]):
+        self.ex_iterables = ex_iterables
+
+    def __iter__(self):
+        for ex_iterable in self.ex_iterables:
+            yield from ex_iterable
+
+    def shuffle_data_sources(
+        self, generator: np.random.Generator
+    ) -> "VerticallyConcatenatedMultiSourcesExamplesIterable":
+        """Shuffle the list of examples iterable, as well as each underlying examples iterable."""
+        rng = deepcopy(generator)
+        ex_iterables = list(self.ex_iterables)
+        rng.shuffle(ex_iterables)
+        ex_iterables = [ex_iterable.shuffle_data_sources(generator) for ex_iterable in ex_iterables]
+        return VerticallyConcatenatedMultiSourcesExamplesIterable(ex_iterables)
+
+    @property
+    def n_shards(self) -> int:
+        return sum(ex_iterable.n_shards for ex_iterable in self.ex_iterables)
+
+    def shard_data_sources(self, shard_idx: int) -> "VerticallyConcatenatedMultiSourcesExamplesIterable":
+        """Either keep only the requested shard, or propagate the request to the underlying iterable."""
+        raise NotImplementedError("Sharding a VerticallyConcatenatedMultiSourcesExamplesIterable is not implemented")
+
+
+def _check_column_names(column_names: List[str]):
+    """Check the column names to make sure they don't contain duplicates."""
+    counter = Counter(column_names)
+    if not all(count == 1 for count in counter.values()):
+        duplicated_columns = [col for col in counter if counter[col] > 1]
+        raise ValueError(
+            f"The examples iterables can't have duplicated columns but columns {duplicated_columns} are duplicated."
+        )
+
+
+class HorizontallyConcatenatedMultiSourcesExamplesIterable(_BaseExamplesIterable):
+    def __init__(self, ex_iterables: List[_BaseExamplesIterable]):
+        self.ex_iterables = ex_iterables
+
+    def __iter__(self):
+        ex_iterators = [iter(ex_iterable) for ex_iterable in self.ex_iterables]
+        for i in itertools.count():
+            keys = []
+            examples = []
+            for ex_iterator in list(ex_iterators):
+                try:
+                    key, example = next(ex_iterator)
+                    keys.append(key)
+                    examples.append(example)
+                except StopIteration:
+                    ex_iterators.remove(ex_iterator)
+            if ex_iterators:
+                if i == 0:
+                    _check_column_names([column_name for example in examples for column_name in example])
```

Comment:
I'd keep it because I think `HorizontallyConcatenatedMultiSourcesExamplesIterable` is also responsible for checking that there are no duplicates. This is also tested at

https://github.com/huggingface/datasets/blob/94b229377df59a0ff6d9fb93cad3c9644fb0a871/tests/test_iterable_dataset.py#L506-L508

Reply:
That's also OK :).